packages/babylon/src/plugins/flow.js

```diff
@@ -2340,6 +2378,103 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       return super.parseSubscripts(base, startPos, startLoc, noCalls);
     }
 
+    parseSubscript(
+      base: N.Expression,
+      startPos: number,
+      startLoc: Position,
+      noCalls: ?boolean,
+      subscriptState: N.ParseSubscriptState,
+    ): N.Expression {
+      if (this.match(tt.questionDot) && this.isLookaheadRelational("<")) {
+        this.expectPlugin("optionalChaining");
+        subscriptState.optionalChainMember = true;
+        if (noCalls) {
+          subscriptState.stop = true;
+          return base;
+        }
+        this.next();
+        const state = this.state.clone();
+        try {
+          const node: N.OptionalCallExpression = this.startNodeAt(
+            startPos,
+            startLoc,
+          );
+          node.callee = base;
+          node.typeArguments = this.flowParseTypeParameterInstantiation();
+          this.expect(tt.parenL);
+          node.arguments = this.parseCallExpressionArguments(tt.parenR, false);
+          node.optional = true;
+          return this.finishNode(node, "OptionalCallExpression");
+        } catch (e) {
+          if (e instanceof SyntaxError) {
+            this.state = state;
+            const node: N.OptionalMemberExpression = this.startNodeAt(
+              startPos,
+              startLoc,
+            );
+            node.object = base;
+            node.property = this.parseIdentifier(true);
```

Comment:
We already know that after `?.` there is `<` because we lookead ahead... When can this be an identifier?

Reply:
You're right. The idea isn't to successfully parse an identifier, but to error as if an identifier were expected if we failed to parse as a call with type args. This is probably overcomplicated, since we could also just call `this.unexpected`. Alternatively, since `o.?<` is not ambiguous, we could avoid the `state.clone` and fallback code entirely. I slightly prefer the latter course. What do you think?

Reply:
OK, pushed an update with this change.